src/zkevm_specs/evm/instruction.py

```diff
@@ -100,6 +101,14 @@ def constrain_bool(self, value: int):
     def constrain_gas_left_not_underflow(self, gas_left: int):
         self.bytes_range_lookup(gas_left, N_BYTES_GAS)
 
+    def constrain_execution_state_transition(self):
+        if self.next.execution_state == ExecutionState.BeginTx:
+            assert self.curr.execution_state == ExecutionState.EndTx
+        elif self.next.execution_state == ExecutionState.EndTx:
+            assert self.curr.execution_state.halts() or self.curr.execution_state == ExecutionState.BeginTx
```

Comment:
what is the case of  the `BeginTx:` direct follows by `EndTx` ?

Reply:
When `callee` has empty code. But we could also remove this branch if we decide to always dive into call, even if code is empty.